File: clients/godot-app/scripts/Hud.cs
using Godot;

namespace Playground
{
    /// <summary>
    /// Immediate-mode side panel: telemetry rows, sparklines, pending-input
    /// chips, and the keyboard "controls" list that stands in for the web
    /// build's sliders and toggles. Port of framework/hud.ts + controls.ts.
    /// </summary>
    public class Hud
    {
        private float _x, _y, _w, _cursor;

        public void Begin(float x, float y, float w)
        {
            _x = x; _y = y; _w = w; _cursor = y;
        }

        public void Section(string title)
        {
            _cursor += 10;
            Draw.Text(_x, _cursor, _w, title, Palette.TextFaint, 10);
            _cursor += 14;
            Draw.Rect(new Rect2(_x, _cursor, _w, 1), Palette.Border);
            _cursor += 6;
        }

        public void Row(string label, string value, Color valueColor)
        {
            Draw.Text(_x, _cursor, _w, label, Palette.TextDim, 12);
            Draw.Text(_x, _cursor, _w, value, valueColor, 12, HorizontalAlignment.Right);
            _cursor += 18;
        }

        public void Key(string key, string what)
        {
            Draw.Rect(new Rect2(_x, _cursor - 1, 58, 15), Palette.Inset);
            Draw.Text(_x + 4, _cursor - 2, 60, key, Palette.Text, 10);
            Draw.Text(_x + 64, _cursor - 2, _w - 64, what, Palette.TextDim, 10);
            _cursor += 17;
        }

        public void Note(string text)
        {
            float h = Draw.TextWrapped(_x, _cursor, _w, text, Palette.TextFaint, 10);
            _cursor += h + 6;
        }

        /// <summary>One chip per pending (unacked) input — drains as acks land.</summary>
        public void Chips(string label, int n, int max = 48)
        {
            Row(label, n.ToString(), Palette.Accent);
            float cx = _x, cy = _cursor;
            for (int i = 0; i < Mathf.Min(n, max); i++)
            {
                Draw.Rect(new Rect2(cx, cy, 5, 10), Palette.A(Palette.Accent, 0.75f));
                cx += 7;
                if (cx + 5 > _x + _w) { cx = _x; cy += 12; }
            }
            _cursor = cy + 18;
        }

        public void Spark(Spark s, string label, string value, Color color, double fixedMax = 0)
        {
            Row(label, value, color);
            const float gh = 30;
            var box = new Rect2(_x, _cursor, _w, gh);
            Draw.Rect(box, Palette.Inset);
            s.Render(box, color, fixedMax);
            _cursor += gh + 8;
        }
    }

    /// <summary>Rolling graph; auto-scales to the window max unless given one.</summary>
    public class Spark
    {
        private const int Cap = 140;
        private readonly double[] _data = new double[Cap];
        private int _head, _count;

        public void Push(double v)
        {
            _data[_head] = v;
            _head = (_head + 1) % Cap;
            if (_count < Cap) _count++;
        }

        public double Last => _count == 0 ? 0 : _data[(_head - 1 + Cap) % Cap];

        public void Render(Rect2 box, Color color, double fixedMax)
        {
            if (_count < 2) return;
            double max = fixedMax;
            if (max <= 0) { for (int i = 0; i < _count; i++) if (_data[i] > max) max = _data[i]; }
            if (max <= 0) max = 1e-6;
            int start = (_head - _count + Cap * 2) % Cap;
            for (int i = 1; i < _count; i++)
            {
                double v0 = _data[(start + i - 1) % Cap], v1 = _data[(start + i) % Cap];
                float x0 = box.Position.X + (i - 1) / (float)(Cap - 1) * box.Size.X;
                float x1 = box.Position.X + i / (float)(Cap - 1) * box.Size.X;
                float y0 = box.End.Y - 2 - (float)System.Math.Min(1, v0 / max) * (box.Size.Y - 4);
                float y1 = box.End.Y - 2 - (float)System.Math.Min(1, v1 / max) * (box.Size.Y - 4);
                Draw.Line(new Vector2(x0, y0), new Vector2(x1, y1), color);
            }
        }
    }
}

File: clients/flutter-app/lib/hud.dart
import 'dart:math' as math;

import 'package:flutter/widgets.dart';

import 'draw_kit.dart';
import 'palette.dart';
import 'spark.dart';

/// Colour bands for a telemetry value — "is this number healthy?".
enum HudTone {
  /// Neutral readout.
  plain(Palette.text),

  /// Within budget.
  good(Palette.good),

  /// Drifting.
  warn(Palette.warn),

  /// Broken.
  bad(Palette.bad);

  const HudTone(this.color);

  /// The value colour this tone paints with.
  final Color color;
}

/// Immediate-mode painted side panel: telemetry rows, sparklines,
/// pending-input chips and the keyboard-controls list.
///
/// Port of `godot-gd-app/scripts/hud.gd` — same cursor arithmetic, same sizes.
/// Call [begin] once per frame, then emit rows top-down; [cursor] is where the
/// next element would land.
class Hud {
  Hud(this.kit);

  /// The kit the HUD paints into. Reassign per frame when the Hud outlives the
  /// canvas (a [DrawKit] is per-`paint()`).
  DrawKit kit;

  double _x = 0.0;
  double _y = 0.0;
  double _w = 0.0;
  double _cursor = 0.0;

  /// Screen y where the next element would be drawn.
  double get cursor => _cursor;

  /// Start a column of [width] px at screen ([x], [y]).
  void begin(double x, double y, double width) {
    _x = x;
    _y = y;
    _w = width;
    _cursor = y;
  }

  /// Underlined group heading.
  void section(String title) {
    _cursor += 10;
    kit.text(_x, _cursor, _w, title, Palette.textFaint, size: 10);
    _cursor += 14;
    kit.rect(Rect.fromLTWH(_x, _cursor, _w, 1), Palette.border);
    _cursor += 6;
  }

  /// Label on the left, value right-aligned and tinted by [tone] — or by
  /// [color], for the rows whose colour identifies a series rather than a
  /// verdict (sparklines, per-lane readouts).
  void row(
    String label,
    String value, {
    HudTone tone = HudTone.plain,
    Color? color,
  }) {
    kit.text(_x, _cursor, _w, label, Palette.textDim, size: 12);
    kit.text(_x, _cursor, _w, value, color ?? tone.color,
        size: 12, align: TextAlign.right);
    _cursor += 18;
  }

  /// A keybinding line: the key in a chip, its meaning beside it. Stands in
  /// for the web build's on-screen controls where a lab is keyboard-only.
  void keyHint(String key, String desc) {
    kit.rect(Rect.fromLTWH(_x, _cursor - 1, 58, 15), Palette.inset);
    kit.text(_x + 4, _cursor - 2, 60, key, Palette.text, size: 10);
    kit.text(_x + 64, _cursor - 2, _w - 64, desc, Palette.textDim, size: 10);
    _cursor += 17;
  }

  /// Wrapped caption.
  void note(String text) {
    final h = kit.textWrapped(_x, _cursor, _w, text, Palette.textFaint, size: 10);
    _cursor += h + 6;
  }

  /// One chip per pending (unacked) input — the row drains as acks land.
  void chips(int count, {String label = 'pending', int max = 48}) {
    row(label, '$count', color: Palette.accent);
    // The count row above is the truth; the chips are the feel of it.
    var cx = _x;
    var cy = _cursor;
    for (var i = 0; i < math.min(count, max); i++) {
      kit.rect(Rect.fromLTWH(cx, cy, 5, 10), Palette.accent.fade(0.75));
      cx += 7;
      if (cx + 5 > _x + _w) {
        cx = _x;
        cy += 12;
      }
    }
    _cursor = cy + 18;
  }

  /// A labelled row plus the sparkline's own inset plot. [value] defaults to
  /// the newest sample at 3 decimals (`hud.ts`'s default format).
  void spark(
    String label,
    Spark s, {
    Color color = Palette.accent,
    String? value,
    double fixedMax = 0,
  }) {
    row(label, value ?? s.last.toStringAsFixed(3), color: color);
    const gh = 30.0;
    final box = Rect.fromLTWH(_x, _cursor, _w, gh);
    kit.rect(box, Palette.inset);
    s.draw(kit, box, color, fixedMax: fixedMax);
    _cursor += gh + 8;
  }

  /// Blank vertical space.
  void gap([double px = 8]) => _cursor += px;

  /// Height consumed since [begin].
  double get height => _cursor - _y;
}
